### PR TITLE
Tensorflow MLPerf tests for D05

### PR DESCRIPTION
**Pull request by K1504296**
Creating automation of Tensorflow MLPerf tests on the D05 : resnet50, ssd-resnet32, and ssd-mobilenet. These scripts can be run with or without mounting the device to a NFS drive with pre downloaded data sets (only available for devices in Cambridge COLO).

These scripts run a Tensorflow Benchmark with the equivalent dataset (resnet50, ssd-mobilenet etc.) and are meant to give accurate benchmark data on the Tensorflow Library.

Signed-off-by: Theodore Grey <theodore.grey@linaro.org>

### CODE REVIEW

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 1]
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+
+# shellcheck disable=SC1091
+
+. ../../lib/sh-test-lib
+OUTPUT="$(pwd)/output"
+
+usage() {
+    echo "Usage: $0 [-a <home-directory>]
+                    [-m <mount-directory>]
```
Comment (nareshkamboju) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:10]: [-t <true|false>] 
Do you need this in usage() ?

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    ck pull repo:ck-env
+    python -m pip install scikit-build
+    python -m pip install --extra-index-url https://snapshots.linaro.org/ldcg/python-cache/ tensorflow-io-gcs-filesystem==0.21.0 h5py==3.1.0
+    python -m pip install tensorflow-aarch64==2.7.0
+    mkdir "${HOME_DIR}"/src
+    pushd "${HOME_DIR}"/src || exit
+    git clone https://github.com/mlcommons/inference.git
+    pushd inference || exit
+    git checkout r1.1
+    git cherry-pick -n 215c057fc6690a47f3f66c72c076a8f73d66cb12
```
Comment (nareshkamboju) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:57]: Usage the get_test_program() for clone and checkout.
Refer the file: automated/lib/sh-test-lib
get_test_program ${TEST_GIT_URL}  \
${TEST_DIR} \
${TEST_PROG_VERSION} \
${TEST_PROGRAM}

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+popd || exit
+
+if [ "${MNT_EXSTS}" = false ]; then
+    
+    mkdir models
+    pushd models || exit
+    ck install package --tags=image-classification,dataset,imagenet,aux
+    ck install package --tags=image-classification,dataset,imagenet,val
+    cp "${HOME_DIR}"/CK-TOOLS/dataset-imagenet-ilsvrc2012-aux-from.dividiti/val.txt "${HOME_DIR}"/CK-TOOLS/dataset-imagenet-ilsvrc2012-val-min/val_map.txt
+    wget https://zenodo.org/record/2535873/files/resnet50_v1.pb
```
Comment (nareshkamboju) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:75]: Recommended to us curl over wget.
curl -sSOL "${RESNET_URL}"

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    case "$o" in
+        a) HOME_DIR="${OPTARG}";;
+        m) MNT_DIR="${OPTARG}" ;;
+        t) MNT_EXSTS="${OPTARG}" ;;
+        s) SKIP_INSTALL="${OPTARG}" ;;
+        *) usage ;;
+    esac
+done
+
+pkgs="build-essential git python3-venv python3-dev libhdf5-dev gcc-aarch64-linux-gnu"
```
Comment (nareshkamboju) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:25]: Add "curl" to pkgs.

**Review by nareshkamboju — CHANGES_REQUESTED** [review]
Thanks for your patch.
I have provided review comments for two files and those same comments applicable to the rest of the files.

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh (hunk 3 above)
Comment (bhcopeland) [reply automated/linux/imagenet-resnet50/imagenet-resnet50.sh:75]: To be fair either or is okay when downloading a file

Comment (K1504296) [reply automated/linux/imagenet-resnet50/imagenet-resnet50.sh:75]: I'll make the change to curl, no worries :)

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh (hunk 1 above)
Comment (K1504296) [reply automated/linux/imagenet-resnet50/imagenet-resnet50.sh:10]: Thanks for catching this, resolved.

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh (hunk 2 above)
Comment (K1504296) [reply automated/linux/imagenet-resnet50/imagenet-resnet50.sh]: Thanks for the info on get_test_program. If you could double check that I've implemented it correctly I'd appreciate it.

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 5]
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+# shellcheck disable=SC1091
+
+. ../../lib/sh-test-lib
+OUTPUT="$(pwd)/output"
+
```
Comment (roxell) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:6]: If you only run this script you need to set these variables too

HOME_DIR='/home/debiand05'
MNT_DIR='/mnt'
MNT_EXISTS=true

if not wont it be messy?

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        t) MNT_EXISTS="${OPTARG}" ;;
+        s) SKIP_INSTALL="${OPTARG}" ;;
+        p) TEST_DIR="${OPTARG}" ;;
+        v) TEST_PROG_VERSION="${OPTARG}" ;;
+        u) TEST_GIT_URL="${OPTARG}" ;;
+        *) usage ;;
+    esac
+done
+
+pkgs="build-essential git python3-venv python3-dev libhdf5-dev gcc-aarch64-linux-gnu curl"
```
Comment (roxell) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:36]: gcc-aarch64-linux-gnu, isn't that what you find when you search on a x86 machine, I don't find that on my aarch64 machine.

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    python -m pip install cython
+    python -m pip install google protobuf==3.20.1
+    python -m pip install --no-binary pycocotools pycocotools
+    python -m pip install absl-py pillow
+    python -m pip install --extra-index-url https://snapshots.linaro.org/ldcg/python-cache/ numpy==1.19.5
+    python -m pip install --extra-index-url https://snapshots.linaro.org/ldcg/python-cache/ matplotlib
+    python -m pip install ck
+    ck pull repo:ck-env
+    python -m pip install scikit-build
+    python -m pip install --extra-index-url https://snapshots.linaro.org/ldcg/python-cache/ tensorflow-io-gcs-filesystem==0.21.0 h5py==3.1.0
```
Comment (roxell) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.sh:61]: isn't most of these steps in the skip_install identical in all the different script files?
would it make sense to have a function for this?

File: automated/linux/imagenet-resnet50/imagenet-resnet50.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    devices:
+        - hip07-d05
+params:
+    HOME_DIR: '/home/debiand05'
+    MNT_DIR: '/mnt'
+    MNT_EXISTS: true
+    SKIP_INSTALL: false
+    TEST_PROG_VERSION: ""
+    TEST_GIT_URL: ""
+    TEST_DIR: ""
```
Comment (roxell) [inline automated/linux/imagenet-resnet50/imagenet-resnet50.yaml:20]: you need to set the variable here too, if no the 'get_test_program()' will return that it assumes its already installed.

**Comment by gctucker** [comment]
> (only available for devices in COLO)

Out of interest, what is COLO?

File: automated/linux/imagenet-resnet50/imagenet-resnet50.sh (hunk 7 above)
Comment (K1504296) [reply automated/linux/imagenet-resnet50/imagenet-resnet50.sh]: You're correct, I think most of the setup steps in the skip_install are identical and the only differences are when getting the dataset in MNT_EXISTS. Would you prefer I created a function for this or is it alright to leave as is?

File: automated/linux/tensorflow/imagenet-resnet50.sh [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        popd  || exit
+    fi
+fi
+
+pushd "${HOME_DIR}"/src/inference/vision/classification_and_detection/ || exit
+python setup.py develop
+
+if [ "${MNT_EXISTS}" = true ]; then
+    mkdir "${MNT_DIR}"/datasets
+    mount -t nfs 10.40.96.10:/mnt/nvme "${MNT_DIR}"
```
[the remark this answers is not in the document]
Comment (K1504296) [reply automated/linux/tensorflow/imagenet-resnet50.sh:76]: I think for lab / testing purposes the IP will remain the same

File: automated/linux/tensorflow/imagenet-resnet50.sh [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    tensorflow_pip_install
+    popd || exit
+    mkdir "${HOME_DIR}"/src
+    pushd "${HOME_DIR}"/src || exit
+    get_test_program "${TEST_GIT_URL}" "${TEST_DIR}" "${TEST_PROG_VERSION}" "${TEST_PROGRAM}"
+    pushd loadgen || exit
+    python setup.py develop
+    popd || exit
+    popd || exit
+    if [ "${MNT_EXISTS}" = false ]; then
```
[the remark this answers is not in the document]
Comment (K1504296) [reply automated/linux/tensorflow/imagenet-resnet50.sh]: I did the same with the if mnt_exists = true

**Comment by nareshkamboju** [comment]
@K1504296 there are few shellcheck error. please resolve them.
https://github.com/Linaro/test-definitions/runs/7114296167?check_suite_focus=true

**Review by nareshkamboju — APPROVED** [review]

**Comment by nareshkamboju** [comment]
@K1504296 please share test run log from manual run or LAVA test job link.

**Comment by K1504296** [comment]
> @K1504296 please share test run log from manual run or LAVA test job link.

I have previous jobs of the tests running here prior to formatting for test-definitions repo.

https://validation.linaro.org/scheduler/job/3479014
https://validation.linaro.org/scheduler/job/3478467
https://validation.linaro.org/scheduler/job/3395208

I will run the updated versions and post a comment with them when the jobs are complete.

**Comment by nareshkamboju** [comment]
@K1504296 Thanks for your patch.
Please fix shellcheck failures.

+ python3 validate.py -r build-error.txt -p E501 W503 E203 -s SC1091 SC2230
* SHELLCHECK: [FAILED]: ./automated/linux/tensorflow/imagenet-resnet50.sh
* SHELLCHECK: [OUTPUT]:
 In ./automated/linux/tensorflow/imagenet-resnet50.sh line [14](https://github.com/Linaro/test-definitions/pull/366/checks#step:5:15):
 SSD_34=false
 ^----^ SC2034: SSD_34 appears unused. Verify use (or export if used externally).
 For more information:
  [ https://www.shellcheck.net/wiki/SC2](https://www.shellcheck.net/wiki/SC2034)034 -- SSD_34 appears unused. Verify use.

**Comment by nareshkamboju** [comment]
@K1504296 I have gone through your LAVA jobs but i do not find a way to capture test results.

https://validation.linaro.org/results/3478467

You may refer other jobs the way it capture the Results / Results format / Benchmark results.
https://lkft.validation.linaro.org/results/5241617
https://lkft.validation.linaro.org/results/5240100/1_ssuite

**Pull request merged**